### ISSUE TITLE
NPC control script – characters getting stun locked into hurt animation and won't move

### ISSUE THREAD

**Issue by mike123yy** [TheGameCreators/GameGuruRepo#5776]
I'm using the NPC control script for a boss fight and when you shoot the character it gets locked into the hurt animation that repeats over and over again so the character won't move. It looks kind of cool and cinematic but it makes for a terrible boss fight. You don't notice this happening if the NPC's health is way down but when you turn it up for a character that has a lot of health it happens constantly. That would probably be because they die before this can happen. I made a video showing what happens here:

https://drive.google.com/file/d/1a2wPzQcYCQdoSpykBHR5tPwnLpG6qBAP/view?usp=sharing

**Comment by Kasseyus** [comment]
Hi @mike123yy 

Can you please send that monster to chris@thegamecreators.com and I will be able to investigate further.

**Comment by mike123yy** [comment]
I just emailed him to you. I should mention I have used the script with a couple different boss fights with monsters and they all do the same thing. Make sure to set the health to probably at least 2000 so he doesn't die too quickly to get the same result.

**Comment by Necrym59** [comment]
Why did you walk off the navmesh then back on the navmesh, its not very clear what is happening, perhaps a clearer video showing the player only and the enemy and the navmesh would help identify the problem

**Comment by mike123yy** [comment]
I was stepping backward to try to get him to glitch out and accidentally stepped into a pond in the level. I didn't know that would make a difference. 
I made a new video, the first half is a blank map and the second half is my level map but I turned on the daylight. Whats interesting is on the blank map it doesn't really glitch out. It does do the hurt animation every time the character is shot but I can understand why that happens, although it would be helpful if you could adjust the frequency of that. On the level with a lot of trees and obstacles it glitches out almost immediately. 
I turned on diagnostics too so you could watch what's happening when it's glitching. My screen recorder was acting funky so there's no audio and some stuff turned purple but I think it gets the point across  :-)

https://drive.google.com/file/d/1YNKk_PT1Jq-gD3n14SkerlAkKyFD3DSO/view?usp=sharing

**Comment by Kasseyus** [comment]
I have managed to (tentatively) reproduce this bug.  Not every time, but it cannot be reproduced in an empty level, but a basic forest biome full of trees will sometimes do it.

As far as I can tell, it's not glitching out on the hurt animation, but flickering between an idle/senses state when the monster looses tracking on the player.

On my example, I engaged him, and he pursued and reacted to bullets normally, until after a short time, reverted to an idle state, even when shot a couple of times.  Then I tried it from behind a tree and got the glitching result.

I'll have to look further into this to try and determine exactly when the issue occurs.

**Comment by Necrym59** [comment]
Hmm maybe its a combination of things - check the view range and angle in the character settings and sense range in the behavior settings as well sometimes getting these right can make a big difference. I will have a play as well and see if there is somthing else.

**Comment by mike123yy** [comment]
I had a different game I was making where it was just the player and the boss and it did the same thing. I changed the weapon the player used to a knife to kinda get around the glitching since the enemy would snap out of it if you got close. But just to check I took the ally out of this level and tried it again and the same thing is still happening. It takes a little longer since he's getting shot less but still happens. My other game didn't have trees but it had other obstacles that made the enemy character get stuck on animations and ignore the player

With all the trees in the way if you aren't constantly shooting at the enemy he does start to immediately ignore you if you aren't in a wide open area. I mess quite a bit with the view range and angle and even turn both of them up to maximum but it doesn't seem to make a difference. If there were a way to get the character to detect you no matter what obstacle was in the way that would help with that issue, possibly the other issue if he's glitching because it's switching between seeing the player and not. May be a tick box option letting the enemy see through obstacles.

**Comment by Necrym59** [comment]
In the init section of the b ehavior thers is a variable
_lostsight[e] = 30_
which controls how long it loses sight of the player before changing to another state. Try changing that value to a higher or lower value and see if that helps, it maybe that a lot of obstcles (eg: trees etc) is obstructing its view. I will attempt to make it more critical of different scenarios when i able.

**Comment by mike123yy** [comment]
Yeah, I tested a bunch of values between zero and 100 and couldn't find any difference. I didn't know how high you could go with the number though. Stepping behind a cluster of trees does make the glitch happen faster

**Comment by Necrym59** [comment]
In your video you are stepping in and out of the navmesh you will have to extend the navmesh to a bigger area with an object (trees dont extend the navmesh)

**Comment by Necrym59** [comment]
You extend the navmesh by putting a static object there for the Navmesh to extend its generated area. It could be a rock or somthing should do it.

**Comment by mike123yy** [comment]
hmm…  I have plenty of static objects surrounding the area like I said, I wonder what could be the problem

**Comment by mike123yy** [comment]
Playing around with this some more I put a character creator character out where the Bigfoot is at and gave him the melee script to see if he would have problems. The characters that use the standard scripts can find you through trees and anywhere else with no problems so the problem does seem isolated to the NPC control script

**Comment by Necrym59** [comment]
Yes but only if your on the Navmesh

**Comment by mike123yy** [comment]
I placed the character creator guy in the exact same spot under the same conditions while running around trying to break it and it always work vs.  The npc script that bugs out very quickly each time so either way the npc controls script seems to have trouble navigating where the standard scripts dont.  (melee/armed/ect)

**Comment by Necrym59** [comment]
I cannot replicate as yet. Your example shows you stepping on/off the navmesh. No character will path to the player off the navmesh. The behaviour does two things, it checks if the player is seen and is on the navemsh if not it gives up the chase. Do a sample map with no trees and show the difference using the same behavior if possible.

**Comment by mike123yy** [comment]
Here's a Cop with a dagger (melee) vs bigfoot (npc script) in the same wooded area 

https://drive.google.com/file/d/1H1nim5NMbMUV-DpLP_qJiWxK6o--LMp2/view?usp=sharing

You can see the cop has no problem navigating but the Bigfoot does and does the glitching

**Comment by Necrym59** [comment]
But you have a third party ally in there as well so this doesnt isolate the problem. PLus comparing two different scripts means nothing in the end.  Send me the Bigfoot character and i will do some tests here to isolate it down. As I said the behaviour does two things, it checks if the player is seen and is on the navemsh, so i will play with isolating which one is giving the exact problem in this circumstance.

**Comment by mike123yy** [comment]
What e-mail should I send the character to?

**Comment by Necrym59** [comment]
necrym@gmail.com
is fine

**Comment by mike123yy** [comment]
It's on this way along with the level I'm using Minus the assets but I could send those if needed to make the level complete and two replicate the navmesh

**Comment by Necrym59** [comment]
I have updated the BiggFoot Model
and have updated the NPC Controller
do you have a direct email i can send them to for you to test

**Comment by mike123yy** [comment]
Here's my e-mail
michealgordon18@yahoo.com

**Comment by mike123yy** [comment]
I gave it a couple play through's with the new script and the updated model you sent and it seems to be working great! He's able to find the player anywhere you run and it no longer glitches out repeating the taunt or hurt animation. I tried a bunch of tactics to ditch him but he kept on chasing, LOL. Works great, thank you!

**Comment by mike123yy** [comment]
well, the issue seems to have come back.  Was there some sort of update to the script or to Max that might have caused it to start happening again.  It was working great for a while though.  I did have a hard drive crash and I was using the script you sent me so maybe the main one is'nt updated ?

**Comment by Necrym59** [comment]
Nothings changed since

**Comment by mike123yy** [comment]
Well, that's strange.  I tested it today and it's back to working without glitching so I have no idea what was happening. I figured it might be the standalone but I just made a new version and it seems to be working there too so…I suppose you could just ignore this post =)
